packages/compile-solidity/src/run.ts

```diff
@@ -70,17 +92,15 @@ async function run(rawSources, options, internalOptions = {}) {
     throw new Common.Errors.CompileError(errors);
   }
 
-  // success case
-  // returns Compilation - see @truffle/compile-common
-  const outputSources = processAllSources({
+  const outputSources: Source[] = processAllSources({
     sources,
     compilerOutput,
     originalSourcePaths,
     language
   });
-  const sourceIndexes = outputSources
+  const sourceIndexes: string[] = outputSources
     ? outputSources.map(source => source.sourcePath)
-    : undefined; //leave undefined if sources undefined
+    : [];
```

Comment:
Sorry that I missed this on my first pass, but I think this change is breaking? That is, assuming that we consider the `run` function part of the public interface of the package.

I'd say it's breaking because code that previously expected an `undefined` for `sourceIndexes` in the return data must now be changed to expect an empty array.

Reply:
I would say that it probably isn't part of the interface. It is not exposed by it. It is exposed, however, in the work done to move the tests [here](https://github.com/trufflesuite/truffle/pull/5445). But that is purely so that it can be tested by this separate packages. What do you think @gnidan? Maybe you have a different opinion.

Reply:
Assuming that the consensus is that these _aren't_ part of the public interface, but that they still need to be exported for testing purposes, I think we should consider using some naming prefix to indicate to would-be consumers that these are not actually part of the public interface.

Historically I've used an underscore or dunder prefix for that, but I'm not really all that fussed about what we actually use - so long as it's communicated externally.

Reply:
Alternatively maybe we just communicate that unless otherwise documented, only the things exported by the module's entrypoint should be considered part of the public interface?

Reply:
> Alternatively maybe we just communicate that unless otherwise documented, only the things exported by the module's entrypoint should be considered part of the public interface? This is probably much easier to enforce in practice.

This is my strong preference - don't force us to use underscores ;)

Particularly following @wbt's excellent concerns raised in https://github.com/trufflesuite/truffle/discussions/5417 (aside: sorry wbt; I know I still owe you a response there), I think we should probably write something similar to how Hardhat has its ["stability guarantees" page](https://hardhat.org/hardhat-runner/docs/reference/stability-guarantees), where it says at the top that they don't "follow semver strictly."

This will probably hurt a bit at first, but it seems best to maintain our public interfaces explicitly as "what is documented is what you can rely on". Sure, sometimes people like you, @wbt, will have to fix breakages as they happen, but that's unavoidable... the only user-facing difference with strict adherence to semver would be that each breakage gets a major release; the work to upgrade remains the same. But it's costly to do major releases, and it's also dangerous for us to let our public interfaces remain so loosely defined as "whatever is exported"... maybe in a few years we'll have the project maturity to support that, but right now this sort of thing keeps causing us to get in our own way.

Reply:
(Oh and please don't fret @wbt; I'm not talking about your recently raised issues about our breaking types. Those are absolutely mea culpa and should've been major releases or not at all. I'm specifically talking here about the part of the interface surface area that we believe is internal)

Reply:
> This is my strong preference - don't force us to use underscores ;)

Ha, I'm not sure that I could even if I wanted to!

Reply:
I'm going to resolve this thread under the assumption that this isn't a change to public interface, and that @gnidan and/or @eggplantzzz will handle figuring out how to communicate our stability guarantees elsewhere.